waiter/src/waiter/scheduler.clj

```diff
@@ -638,3 +642,145 @@
           "WAITER_PASSWORD" (service-id->password-fn service-id)
           "WAITER_SERVICE_ID" service-id
           "WAITER_USERNAME" "waiter"}))
+
+;; Support for tracking service instance launching time stats
+
+(def instance-counts-zero {:requested 0 :started 0})
+
+(defn- make-launch-tracker-entry
+  [service-id]
+  [service-id
+   {:instance-counts instance-counts-zero
+    :known-instance-ids #{}
+    :instance-scheduling-start-times []
+    :service-schedule-timer (metrics/service-timer service-id "launch-overhead" "schedule-time")
+    :service-startup-timer (metrics/service-timer service-id "launch-overhead" "startup-time")
+    :starting-instance-ids []}])
+
+(defn update-launch-trackers
+  "Track metrics on app instances, specifically schedule and startup times."
+  [service-id->launch-tracker new-service-ids removed-service-ids service-id->healthy-instances
+   service-id->unhealthy-instances service-id->instance-counts waiter-schedule-timer]
+  (let [now (t/now)
+        new-tracker-entries (map make-launch-tracker-entry new-service-ids)
+        tracker-entries' (->> service-id->launch-tracker
+                              ;; Remove trackers from deleted services
+                              (remove (comp removed-service-ids key))
```

Comment:
This could be:
```
new-tracker-entries (pc/map-from-keys make-launch-tracker-entry new-service-ids)
tracker-entries' (-> (apply dissoc service-id->launch-tracker removed-service-ids)
                     (merge new-tracker-entries))
```
and then we could actually call this `service-id->launch-tracker'`

Reply:
Updated these. Pretty much the same as what you suggested.